File: apps/flutter/lib/core/models/ble_service.dart
/// BLE 服务模型
class BleService {
  /// 服务 UUID
  final String uuid;

  /// 是否为主服务
  final bool isPrimary;

  /// 服务名称（标准服务会显示名称）
  final String? name;

  /// 特征值列表
  final List<BleCharacteristic> characteristics;

  /// 是否展开（UI 状态）
  bool expanded;

  BleService({
    required this.uuid,
    this.isPrimary = false,
    this.name,
    this.characteristics = const [],
    this.expanded = false,
  });

  /// 显示名称
  String get displayName => name ?? '未知服务';

  /// 短 UUID（去掉连字符并取前 8 位）
  String get shortUuid {
    final clean = uuid.replaceAll('-', '').toUpperCase();
    if (clean.length == 32) {
      // 标准 128 位 UUID
      if (clean.startsWith('0000')) {
        return clean.substring(4, 8);
      }
      return clean.substring(0, 8);
    }
    return clean;
  }

  /// 复制并修改部分属性
  BleService copyWith({
    String? uuid,
    bool? isPrimary,
    String? name,
    List<BleCharacteristic>? characteristics,
    bool? expanded,
  }) {
    return BleService(
      uuid: uuid ?? this.uuid,
      isPrimary: isPrimary ?? this.isPrimary,
      name: name ?? this.name,
      characteristics: characteristics ?? this.characteristics,
      expanded: expanded ?? this.expanded,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BleService && runtimeType == other.runtimeType && uuid == other.uuid;

  @override
  int get hashCode => uuid.hashCode;

  @override
  String toString() {
    return 'BleService(uuid: $uuid, isPrimary: $isPrimary, characteristics: ${characteristics.length})';
  }
}

/// BLE 特征值属性
enum BleCharacteristicProperty {
  /// 可读
  read,
  /// 可写（有响应）
  write,
  /// 可写（无响应）
  writeNoResponse,
  /// 可通知
  notify,
  /// 可指示
  indicate,
}

/// BLE 特征值模型
class BleCharacteristic {
  /// 特征值 UUID
  final String uuid;

  /// 所属服务 UUID
  final String serviceUuid;

  /// 属性列表
  final List<BleCharacteristicProperty> properties;

  /// 当前值
  final List<int>? value;

  /// 是否正在监听通知
  bool isNotifying;

  /// 特征值名称（标准特征值会显示名称）
  final String? name;

  BleCharacteristic({
    required this.uuid,
    required this.serviceUuid,
    this.properties = const [],
    this.value,
    this.isNotifying = false,
    this.name,
  });

  /// 显示名称
  String get displayName => name ?? '未知特征值';

  /// 短 UUID
  String get shortUuid {
    final clean = uuid.replaceAll('-', '').toUpperCase();
    if (clean.length == 32) {
      if (clean.startsWith('0000')) {
        return clean.substring(4, 8);
      }
      return clean.substring(0, 8);
    }
    return clean;
  }

  /// 是否可读
  bool get canRead => properties.contains(BleCharacteristicProperty.read);

  /// 是否可写
  bool get canWrite =>
      properties.contains(BleCharacteristicProperty.write) ||
      properties.contains(BleCharacteristicProperty.writeNoResponse);

  /// 是否可通知
  bool get canNotify =>
      properties.contains(BleCharacteristicProperty.notify) ||
      properties.contains(BleCharacteristicProperty.indicate);

  /// 复制并修改部分属性
  BleCharacteristic copyWith({
    String? uuid,
    String? serviceUuid,
    List<BleCharacteristicProperty>? properties,
    List<int>? value,
    bool? isNotifying,
    String? name,
  }) {
    return BleCharacteristic(
      uuid: uuid ?? this.uuid,
      serviceUuid: serviceUuid ?? this.serviceUuid,
      properties: properties ?? this.properties,
      value: value ?? this.value,
      isNotifying: isNotifying ?? this.isNotifying,
      name: name ?? this.name,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BleCharacteristic &&
          runtimeType == other.runtimeType &&
          uuid == other.uuid &&
          serviceUuid == other.serviceUuid;

  @override
  int get hashCode => uuid.hashCode ^ serviceUuid.hashCode;

  @override
  String toString() {
    return 'BleCharacteristic(uuid: $uuid, properties: ${properties.length}, isNotifying: $isNotifying)';
  }
}

File: apps/android/app/src/main/java/com/smartble/core/model/BleService.kt
package com.smartble.core.model

/**
 * BLE 服务
 */
data class BleService(
    val uuid: String,
    val characteristics: List<BleCharacteristic> = emptyList(),
) {
    val shortUuid: String
        get() = uuid.shortUuid()

    val displayName: String
        get() = BleUuids.getServiceName(uuid)
}

/**
 * BLE 特征值
 */
data class BleCharacteristic(
    val serviceUuid: String,
    val uuid: String,
    val properties: Set<Property>,
    val value: ByteArray? = null,
    val isNotifying: Boolean = false,
) {
    val shortUuid: String
        get() = uuid.shortUuid()

    val displayName: String
        get() = BleUuids.getCharacteristicName(uuid)

    val canRead: Boolean
        get() = properties.contains(Property.Read)

    val canWrite: Boolean
        get() = properties.contains(Property.Write) ||
                 properties.contains(Property.WriteNoResponse)

    val canNotify: Boolean
        get() = properties.contains(Property.Notify) ||
                 properties.contains(Property.Indicate)

    fun copyWithValue(value: ByteArray?): BleCharacteristic {
        return copy(value = value)
    }

    fun copyWithNotifying(notifying: Boolean): BleCharacteristic {
        return copy(isNotifying = notifying)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as BleCharacteristic

        if (serviceUuid != other.serviceUuid) return false
        if (uuid != other.uuid) return false
        if (properties != other.properties) return false
        if (value != null) {
            if (other.value == null) return false
            if (!value.contentEquals(other.value)) return false
        } else if (other.value != null) return false
        if (isNotifying != other.isNotifying) return false

        return true
    }

    override fun hashCode(): Int {
        var result = serviceUuid.hashCode()
        result = 31 * result + uuid.hashCode()
        result = 31 * result + properties.hashCode()
        result = 31 * result + (value?.contentHashCode() ?: 0)
        result = 31 * result + isNotifying.hashCode()
        return result
    }
}

/**
 * 特征值属性
 */
enum class Property {
    Read,
    Write,
    WriteNoResponse,
    Notify,
    Indicate,
    AuthenticatedSignedWrites,
    ExtendedProperties,
}

/**
 * BLE UUID 常量和工具
 */
object BleUuids {
    // 通用服务 UUID
    const val SERVICE_GENERIC_ACCESS = "00001800-0000-1000-8000-00805F9B34FB"
    const val SERVICE_GENERIC_ATTRIBUTE = "00001801-0000-1000-8000-00805F9B34FB"
    const val SERVICE_DEVICE_INFORMATION = "0000180A-0000-1000-8000-00805F9B34FB"
    const val SERVICE_BATTERY = "0000180F-0000-1000-8000-00805F9B34FB"
    const val SERVICE_HUMAN_INTERFACE_DEVICE = "00001812-0000-1000-8000-00805F9B34FB"
    const val SERVICE_OTA = "4fafc201-1fb5-459e-8fcc-c5c9c331914d"

    // 通用特征值 UUID
    const val CHARACTERISTIC_DEVICE_NAME = "00002A00-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_APPEARANCE = "00002A01-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_PERIPHERAL_PRIVACY_FLAG = "00002A02-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_RECONNECTION_ADDRESS = "00002A03-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_PERIPHERAL_PREFERRED_CONNECTION_PARAMETERS = "00002A04-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_SERVICE_CHANGED = "00002A05-0000-1000-8000-00805F9B34FB"

    // 设备信息服务特征值
    const val CHARACTERISTIC_MANUFACTURER_NAME = "00002A29-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_MODEL_NUMBER = "00002A24-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_SERIAL_NUMBER = "00002A25-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_HARDWARE_REVISION = "00002A27-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_FIRMWARE_REVISION = "00002A26-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_SOFTWARE_REVISION = "00002A28-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_SYSTEM_ID = "00002A23-0000-1000-8000-00805F9B34FB"

    // 电池服务特征值
    const val CHARACTERISTIC_BATTERY_LEVEL = "00002A19-0000-1000-8000-00805F9B34FB"

    // OTA 特征值
    const val CHARACTERISTIC_OTA_CONTROL = "beb5483e-36e1-4688-b7f5-ea07361b26c0"
    const val CHARACTERISTIC_OTA_DATA = "beb5483e-36e1-4688-b7f5-ea07361b26c1"
    const val CHARACTERISTIC_OTA_STATUS = "beb5483e-36e1-4688-b7f5-ea07361b26c2"

    private val serviceNames = mapOf(
        SERVICE_GENERIC_ACCESS to "Generic Access",
        SERVICE_GENERIC_ATTRIBUTE to "Generic Attribute",
        SERVICE_DEVICE_INFORMATION to "Device Information",
        SERVICE_BATTERY to "Battery Service",
        SERVICE_HUMAN_INTERFACE_DEVICE to "HID",
        SERVICE_OTA to "OTA Service",
    )

    private val characteristicNames = mapOf(
        CHARACTERISTIC_DEVICE_NAME to "Device Name",
        CHARACTERISTIC_APPEARANCE to "Appearance",
        CHARACTERISTIC_PERIPHERAL_PRIVACY_FLAG to "Privacy Flag",
        CHARACTERISTIC_RECONNECTION_ADDRESS to "Reconnection Address",
        CHARACTERISTIC_PERIPHERAL_PREFERRED_CONNECTION_PARAMETERS to "Connection Parameters",
        CHARACTERISTIC_SERVICE_CHANGED to "Service Changed",
        CHARACTERISTIC_MANUFACTURER_NAME to "Manufacturer Name",
        CHARACTERISTIC_MODEL_NUMBER to "Model Number",
        CHARACTERISTIC_SERIAL_NUMBER to "Serial Number",
        CHARACTERISTIC_HARDWARE_REVISION to "Hardware Revision",
        CHARACTERISTIC_FIRMWARE_REVISION to "Firmware Revision",
        CHARACTERISTIC_SOFTWARE_REVISION to "Software Revision",
        CHARACTERISTIC_SYSTEM_ID to "System ID",
        CHARACTERISTIC_BATTERY_LEVEL to "Battery Level",
        CHARACTERISTIC_OTA_CONTROL to "OTA Control",
        CHARACTERISTIC_OTA_DATA to "OTA Data",
        CHARACTERISTIC_OTA_STATUS to "OTA Status",
    )

    fun getServiceName(uuid: String): String {
        return serviceNames[uuid] ?: "Unknown Service"
    }

    fun getCharacteristicName(uuid: String): String {
        return characteristicNames[uuid] ?: "Unknown Characteristic"
    }
}

/**
 * UUID 扩展函数
 */
fun String.shortUuid(): String {
    return if (length == 36) {
        substring(4, 8)
    } else {
        this
    }
}


